### PR TITLE
Include assembly and file versions in deps file for non self-contained publish

### LINKED ISSUE #1847: Add additional metadata to deps.json to support "framework wins over app"
This is the SDK work to support the core-setup feature https://github.com/dotnet/core-setup/issues/3546

That core-setup issue adds new metadata to the framework's deps.json to help determine which assemblies are newer during file-based probing. This would need to be done by the SDK during build and\or publish. That metadata includes assembly version and file version, where assembly version is compared first and if equal, then file version is compared.

@nguerrera

### LINKED ISSUE #1942: deps file is not regenerated when non-nuget references change
Currently the target that generates the deps file only runs when project.assets.json is older than deps.json, but if non-package references are changed, then the deps file would also need to change.

We do want to keep incrementality because it is an expensive operation, but we should account for the other task parameters that can change the deps file.

### PR DESCRIPTION
**Pull request by dsplaisted**
Fixes #1847

Also fixes #1942 (in most cases)

### CODE REVIEW

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs [hunk 1]
```diff
@@ -132,7 +134,13 @@ protected override void ExecuteCore()
                 PlatformLibraryName,
                 IsSelfContained);
 
-            DependencyContext dependencyContext = new DependencyContextBuilder(mainProject, projectContext)
+            //  If publishing a framework-dependent app, then include the assembly and file versions of files in the
+            //  deps.json to support rolling forward to a newer shared framework.  See https://github.com/dotnet/core-setup/pull/3704
+            //  Reading these versions could impact build perf, so we avoid doing it if we aren't publishing or if we are
+            //  publishing a self-contained app.
+            bool includeRuntimeFileVersions = IsPublish && !IsSelfContained;
```
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs]: It may be better to use `projectContext.IsFrameworkDependent` here instead of `!IsSelfContained`.  The comment even says "publishing a framework-dependent app".

Comment (eerhardt) [reply src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs]: What do you think about moving this logic out of the C# task all together and making a property for "IncludeRuntimeFileVersions" that is defaulted in the MSBuild XML?  That way someone could turn this on even if they weren't publishing, or in a self-contained app in case they wanted this info for some reason.

**Comment by eerhardt** [comment]
> Can you help me figure out why the test is failing?

Can you try repro'ing it outside of the test, and then set `COREHOST_TRACE=1` and inspect the output?  That should tell you why the host choose what it did.

Also, ensure you are using a recent host build.  How old of a version are you using?

**Comment by dsplaisted** [comment]
@eerhardt Here's the COREHOST_TRACE output:

[corehost_trace.txt](https://github.com/dotnet/sdk/files/1885158/corehost_trace.txt)

I think the version of the host I'm using is 2.1.0-preview2-26314-02 (that's the version of the runtime that's being used).

**Comment by steveharter** [comment]
Here's the log for the two deps entries:
```
Adding runtime asset runtimes/win-x64/lib/netcoreapp2.1/System.Collections.Immutable.dll assemblyVersion= fileVersion= from runtime.win-x64.Microsoft.NETCore.App/2.1.0-preview2-26314-02

Adding runtime asset lib/netstandard2.0/System.Collections.Immutable.dll assemblyVersion=1.2.3.0 fileVersion=4.6.26216.2 from System.Collections.Immutable/1.5.0-preview1-26216-02
```

So it looks like the framework's deps file doesn't have the metadata added yet... So it will never "win" in a conflict.

File: src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs (hunk 1 above)
Comment (nguerrera) [reply src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs]: I think we want to be able to have the info on non-publish optionally.

What Steve and I had considered was:

Default experience:
* Build (F5) -> runtimeconfig.dev.json disables roll forward
* Publish -> writes the info

But there could be an opt-in to have build write the info (taking a perf hit) and not disable roll forward.

**Comment by dsplaisted** [comment]
@eerhardt @steveharter @weshaggard I think this is ready for review.  Per discussion with @eerhardt, the runtime's deps.json is generated using the SDK, so in order to include the file and assembly versions in it, we will need to get this change into the SDK and then update the runtime deps generation process to use the updated SDK.  I've filed #2124 to enable the failing end-to-end test from this PR once that happens.

File: src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets [hunk 2]
```diff
@@ -504,6 +504,13 @@ Copyright (c) .NET Foundation. All rights reserved.
 
     <PropertyGroup>
       <PublishDepsFilePath Condition=" '$(PublishDepsFilePath)' == '' ">$(PublishDir)$(ProjectDepsFileName)</PublishDepsFilePath>
+      <!--
+      If publishing a framework-dependent app, then include the assembly and file versions of files in the
+      deps.json to support rolling forward to a newer shared framework.  See https://github.com/dotnet/core-setup/pull/3704
+      Reading these versions could impact build perf, so we avoid doing it by default if we aren't publishing or if we are
```
Comment (steveharter) [inline src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.Publish.targets]: Have you measured the overhead?

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatAPublishedDepsJsonShouldContainVersionInformation.cs [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+            allRuntimeFiles.Should().NotBeEmpty();
+
+            foreach (var runtimeFile in allRuntimeFiles)
+            {
+                runtimeFile.AssemblyVersion.Should().BeNull();
+                runtimeFile.FileVersion.Should().BeNull();
+            }
+        }
+
+        [Fact(Skip = "Host deps.json doesn't have runtime file version info yet: https://github.com/dotnet/sdk/issues/2124")]
```
Comment (steveharter) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatAPublishedDepsJsonShouldContainVersionInformation.cs]: Did you try mocking the metadata (to see if it will pass when the metadata is available)?

**Review by steveharter — APPROVED** [review]
LGTM pending test success

Do you have an issue or reminder to enable the test once the framework metadata is available?

**Comment by eerhardt** [comment]
> the runtime's deps.json is generated using the SDK, so in order to include the file and assembly versions in it, we will need to get this change into the SDK and then update the runtime deps generation process to use the updated SDK.

I think this is going to cause issues with source-build.  Source-build is currently using `2.1.101` of the SDK to build all of the `.NET Core 2.1` release.  Since these new DependencyModel APIs are only in `.NET Core 2.1`, and the new `2.1.300` SDK will be needed to generate the shared framework's .deps.json file, it will mean that source-build will have to use the latest SDK in order to get this new info generated in the shared framework's deps.json file.

/cc @dseefeld @weshaggard

File: src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs [hunk 4]
```diff
@@ -538,7 +569,7 @@ private IEnumerable<RuntimeLibrary> GetDirectReferenceRuntimeLibraries()
                     name: GetReferenceLibraryName(r),
                     version: r.Version,
                     hash: string.Empty,
-                    runtimeAssemblyGroups: new[] { new RuntimeAssetGroup(string.Empty, r.FileName) },
+                    runtimeAssemblyGroups: new[] { new RuntimeAssetGroup(string.Empty, new[] { CreateRuntimeFile(r.FileName, r.FullPath) }) },
```
Comment (eerhardt) [inline src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs]: Note - when @tannergooding's change #2090 goes in, we will need to merge this with that, so the new references get `CreateRuntimeFile` called on them as well.

**Comment by eerhardt** [comment]
> using System;

(nit) copyright header

---
Refers to: src/Tests/Microsoft.NET.Publish.Tests/GivenThatAPublishedDepsJsonShouldContainVersionInformation.cs:1 in e18c743. [](commit_id = e18c743c8f3ee25c01e6606ad859b9e0415507b1, deletion_comment = False)

File: src/Tests/Microsoft.NET.Restore.Tests/GivenThatWeWantToRestoreProjectsUsingNuGetConfigProperties.cs [hunk 5]
```diff
@@ -156,5 +159,52 @@ private void ExtractNupkg(string nugetCache, string nupkg)
                 }
             }
         }
+
+        private class MockPackageDownloader : IPackageDownloader
```
Comment (eerhardt) [inline src/Tests/Microsoft.NET.Restore.Tests/GivenThatWeWantToRestoreProjectsUsingNuGetConfigProperties.cs]: This seems unrelated.

Comment (dsplaisted) [reply src/Tests/Microsoft.NET.Restore.Tests/GivenThatWeWantToRestoreProjectsUsingNuGetConfigProperties.cs]: This is unrelated to the feature.  I retargeted this to master, and tests targeting .NET Core 2.1 were failing.  So I'm trying to update the stage 0 CLI, which brought in an update to NuGet, which had API changes which required this addition... 🤷‍♂️

Comment (dsplaisted) [inline src/Tests/Microsoft.NET.Restore.Tests/GivenThatWeWantToRestoreProjectsUsingNuGetConfigProperties.cs]: @nkolev92 @rohit21agrawal Can you provide guidance on how to implement an `IPackageDownloader` for a test like this?  I tried naively just implementing one method, and it looks like it's failing when it calls the `SignedPackageReader` getter.  What actually needs to be implemented here?

Comment (rohit21agrawal) [reply src/Tests/Microsoft.NET.Restore.Tests/GivenThatWeWantToRestoreProjectsUsingNuGetConfigProperties.cs]: we have a test implementation here: https://github.com/NuGet/NuGet.Client/blob/dev/test/NuGet.Core.Tests/NuGet.Packaging.Test/NugetPackageUtilTests.cs#L805

Hope that helps!

**Comment by dsplaisted** [comment]
@dotnet-bot test this

**Comment by dsplaisted** [comment]
@dotnet-bot test this

**Comment by dsplaisted** [comment]
@dotnet-bot test this

**Comment by mmitche** [comment]
@dotnet-bot test this please

**Comment by dsplaisted** [comment]
@dotnet-bot test this

**Comment by dsplaisted** [comment]
@dotnet-bot test this please

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatAPublishedDepsJsonShouldContainVersionInformation.cs [hunk 6]
```diff
@@ -255,8 +255,10 @@ string GetRollForwardNetCoreAppVersion()
                 IsSdkProject = true,
                 IsExe = true
             };
-            var testAsset = _testAssetsManager.CreateTestProject(testProject)
-                .Restore(Log, testProject.Name);
+            var testAsset = _testAssetsManager.CreateTestProject(testProject);
+            NuGetConfigWriter.Write(testAsset.TestRoot, NuGetConfigWriter.DotnetCoreMyGetFeed);
```
Comment (dsplaisted) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatAPublishedDepsJsonShouldContainVersionInformation.cs]: @johnbeisner Without adding this feed I get an error on CI: `Unable to find package Microsoft.NETCore.App with version (>= 2.1.0-preview3-26411-06`

Should this package not be in the dotnet-core blob storage feed?

**Review by eerhardt — APPROVED** [review]

File: src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs (hunk 4 above)
Comment (tannergooding) [reply src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs]: #2090 was merged, was this handled?

Comment (dsplaisted) [reply src/Tasks/Microsoft.NET.Build.Tasks/DependencyContextBuilder.cs]: @tannergooding @eerhardt It looks like this is handled correctly, as the new code from #2090 ends up calling through `GetReferenceRuntimeLibraries` for the indirect references, the same as how the direct references are handled.

**Comment by dsplaisted** [comment]
> Have you measured the overhead?

@steveharter I've now tested this with an ASP.NET MVC app targeting .NET Core 2.0 (since that will actually write versions to the deps.json, as opposed to 2.1 where all the references are part of the framework and hence won't be in the deps.json).  It shows almost a 5% slowdown on a clean build.  (In an incremental build, the difference is negligible, since the deps.json is considered up to date and not regenerated).

 Scenario Name         | Test Name          | Metric         | Unit | Iterations |  Average | STDEV.S |      Min |      Max
:--------------------- |:------------------ |:-------------- |:----:|:----------:| --------:| -------:| --------:| --------:
 MVC app               | Build              | Execution Time |  ms  |     9      | 2754.570 |  36.997 | 2710.931 | 2812.722
 MVC app               | Build (no changes) | Execution Time |  ms  |     9      | 1302.705 |  12.512 | 1282.533 | 1316.914
 MVC app with versions | Build              | Execution Time |  ms  |     9      | 2884.598 |  21.924 | 2857.216 | 2930.733
 MVC app with versions | Build (no changes) | Execution Time |  ms  |     9      | 1311.112 |  22.929 | 1279.149 | 1349.088

So I think it makes sense to leave these versions out of the deps file by default on build.

**Comment by ahsonkhan** [comment]
@steveharter, @dsplaisted, when will this change get merged? It is blocking release due to the following issue: https://github.com/dotnet/corefx/issues/29032

**Comment by ahsonkhan** [comment]
cc @joshfree, @weshaggard, @ericstj

**Comment by dleeapho** [comment]
> I think this is going to cause issues with source-build. Source-build is currently using 2.1.101 of the SDK to build all of the .NET Core 2.1 release. Since these new DependencyModel APIs are only in .NET Core 2.1, and the new 2.1.300 SDK will be needed to generate the shared framework's .deps.json file, it will mean that source-build will have to use the latest SDK in order to get this new info generated in the shared framework's deps.json file.

source-build is using 2.1.101 SDK to build 2.1 Preview 2 (producing a source-built 2.1.300 SDK). We will then move to using 2.1.300 SDK to build RC. We need to do this because msbuild also needs 2.1.300 SDK to build. I think source-build is good here.

**Comment by dleeapho** [comment]
@dsplaisted no, RC is before RTM for source-build. The point is that we are bootstrapping the use of 2.1.300 SDK as a source-build toolset for RC, RTW via Preview 2.

**Comment by dsplaisted** [comment]
@dotnet-bot test OSX10.12 Release

File: src/Tasks/Microsoft.NET.Build.Tasks/NuGetPackageResolver.cs [hunk 7]
```diff
@@ -41,6 +41,15 @@ public string GetPackageDirectory(string packageId, NuGetVersion version, out st
...
         }
 
+        public string ResolvePackageAssetPath(LockFileTargetLibrary package, string relativePath)
+        {
+            string packagePath = GetPackageDirectory(package.Name, package.Version);
+            return Path.Combine(packagePath, NormalizeRelativePath(relativePath));
+        }
+
+        public static string NormalizeRelativePath(string relativePath)
+                => relativePath.Replace('/', Path.DirectorySeparatorChar);
```
Comment (nguerrera) [inline src/Tasks/Microsoft.NET.Build.Tasks/NuGetPackageResolver.cs]: nit: indentation looks non-standard here.

**Review by nguerrera — APPROVED** [review]

**Comment by dsplaisted** [comment]
@ahsonkhan It just got merged

**Comment by ahsonkhan** [comment]
@dsplaisted, what version of the cli/sdk do we need to test that this change fixed https://github.com/dotnet/corefx/issues/29032?

**Comment by dsplaisted** [comment]
@mmitche can you tell us when there's a prodcon build with this change?

**Comment by nguerrera** [comment]
Make note again that a follow-up change is needed to get versions in to runtime's deps.json as well.

**Comment by ericstj** [comment]
Different error is normal, its failing earlier now: on load of the wrong assembly, rather than later on when it tries to find API that isn't there.
